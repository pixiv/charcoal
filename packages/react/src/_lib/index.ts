/**
 * 今後ポートされる予定の汎用的な関数群
 */

/**
 * Function used to assert a given code path is unreachable
 */
export function unreachable(): never
/**
 * Function used to assert a given code path is unreachable.
 * Very useful for ensuring switches are exhaustive:
 *
 * ```ts
 * switch (a.type) {
 *   case Types.A:
 *   case Types.B:
 *     break
 *   default:
 *     unreachable(a) // will cause a build error if there was
 *                    // a Types.C that was not checked
 * }
 * ```
 *
 * @param value Value to be asserted as unreachable
 */
// NOTE: Uses separate overloads, _not_ `value?: never`, to not allow `undefined` to be passed

export function unreachable(value: never): never
export function unreachable(value?: never): never {
  throw new Error(
    arguments.length === 0
      ? 'unreachable'
      : `unreachable (${JSON.stringify(value)})`
  )
}

/**
 * 複数のrefをマージする。
 *
 * forwardRefで受け取ったrefと、コンポーネント内で定義したrefを同じ要素につけたいケースなどで使う
 */
export function mergeRefs<T>(...refs: React.Ref<T>[]): React.RefCallback<T> {
  return (value) => {
    for (const ref of refs) {
      if (typeof ref === 'function') {
        ref(value)
      } else if (ref !== null) {
        // TODO: use RefObject
        // https://react.dev/blog/2024/04/25/react-19-upgrade-guide#:~:text=%3D%201%3B-,MutableRef,-is%20now%20deprecated
        ;(ref as React.MutableRefObject<T | null>).current = value
      }
    }
  }
}

export function countCodePointsInString(string: string): number {
  // [...string] とするとproduction buildで動かなくなる
  // cf. https://twitter.com/f_subal/status/1497214727511891972
  return Array.from(string).length
}
