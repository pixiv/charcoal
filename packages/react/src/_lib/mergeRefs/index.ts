import React from 'react'

/**
 * Custom implementation of mergeRefs to replace @react-aria/utils
 * Merges multiple refs into a single ref callback
 */
export function mergeRefs<T>(
  ...refs: (React.Ref<T> | undefined)[]
): React.RefCallback<T> {
  return (value: T | null) => {
    refs.forEach((ref) => {
      if (ref == null) {
        return
      }

      if (typeof ref === 'function') {
        ref(value)
      } else if (typeof ref === 'object') {
        ;(ref as React.MutableRefObject<T | null>).current = value
      }
    })
  }
}
