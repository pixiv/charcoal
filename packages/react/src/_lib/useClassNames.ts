import { useMemo } from 'react'

/**
 * Join some class names if propsClassName is defined.
 */
export function useClassNames(...classNames: (string | undefined)[]): string {
  return useMemo(() => classNames.filter((v) => v).join(' '), [classNames])
}
