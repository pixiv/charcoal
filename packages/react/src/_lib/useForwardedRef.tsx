import * as React from 'react'

export function useForwardedRef<T>(ref: React.ForwardedRef<T>): React.RefObject<T> {
  const innerRef = React.useRef<T>(null)

  React.useEffect(() => {
    if (!ref) return
    if (typeof ref === 'function') {
      ref(innerRef.current)
    } else {
      ref.current = innerRef.current
    }
  })

  return innerRef
}
