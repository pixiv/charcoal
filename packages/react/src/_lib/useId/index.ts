import { useId as useReactId } from 'react'

/**
 * Custom useId hook that accepts an optional fallback id
 * Returns the provided id if given, otherwise generates a unique id using React's useId
 */
export function useId(id?: string): string {
  const generatedId = useReactId()
  return id ?? generatedId
}
