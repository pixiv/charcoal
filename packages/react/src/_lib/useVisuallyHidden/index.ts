import './visuallyHidden.css'

/**
 * Custom implementation of useVisuallyHidden to replace @react-aria/visually-hidden
 * Returns props to make an element visually hidden but accessible to screen readers
 */
export function useVisuallyHidden(): {
  visuallyHiddenProps: { className: string }
} {
  return {
    visuallyHiddenProps: {
      className: 'charcoal-visually-hidden',
    },
  }
}
