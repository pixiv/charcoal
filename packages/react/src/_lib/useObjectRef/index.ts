import {
  useCallback,
  useMemo,
  useRef,
  type MutableRefObject,
  type Ref,
} from 'react'
import { mergeRefs } from '../mergeRefs'

/**
 * Custom implementation of useObjectRef to replace @react-aria/utils
 * Converts any ref to a mutable ref object
 */
export function useObjectRef<T>(
  forwardedRef?: Ref<T>
): MutableRefObject<T | null> {
  const objRef = useRef<T | null>(null)
  const _ref = useMemo(() => mergeRefs(objRef, forwardedRef), [forwardedRef])

  const ref = useCallback((value: T | null) => _ref(value), [_ref])

  // Return an object that can be used as both a callback ref and a mutable ref
  return {
    get current() {
      return objRef.current
    },
    set current(value: T | null) {
      ref(value)
    },
  }
}
