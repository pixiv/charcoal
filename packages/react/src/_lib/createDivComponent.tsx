import { DetailedHTMLProps, DO_NOT_USE_OR_YOU_WILL_BE_FIRED_CALLBACK_REF_RETURN_VALUES, forwardRef, ForwardRefExoticComponent, HTMLAttributes, RefAttributes, RefObject } from 'react'
import { useClassNames } from './useClassNames'

export function createDivComponent(mainClassName: string): ForwardRefExoticComponent<Omit<Omit<DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>, "ref"> & {
  ref?: ((instance: HTMLDivElement | null) => void | DO_NOT_USE_OR_YOU_WILL_BE_FIRED_CALLBACK_REF_RETURN_VALUES[keyof DO_NOT_USE_OR_YOU_WILL_BE_FIRED_CALLBACK_REF_RETURN_VALUES]) | RefObject<HTMLDivElement> | null | undefined
}, "ref"> & RefAttributes<HTMLDivElement>> {
  return forwardRef<HTMLDivElement, React.ComponentPropsWithRef<'div'>>(
    function DivComponent({ className, ...props }, ref) {
      const classNames = useClassNames(mainClassName, className)
      return <div className={classNames} ref={ref} {...props} />
    }
  )
}
