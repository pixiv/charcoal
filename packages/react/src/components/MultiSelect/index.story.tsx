import React, { useState } from 'react'
import { Story } from '../../_lib/compat'
import styled from 'styled-components'
import { MultiSelectGroup, default as MultiSelect } from '.'

export default {
  title: 'MultiSelect',
  component: MultiSelect,
  argTypes: {
    name: {
      control: {
        type: 'text',
      },
    },
    ariaLabel: {
      control: {
        type: 'text',
      },
    },
    selected: {
      control: {
        type: 'boolean',
      },
    },
    firstOptionForceChecked: {
      control: {
        type: 'boolean',
      },
    },
    disabled: {
      control: {
        type: 'boolean',
      },
    },
    readonly: {
      control: {
        type: 'boolean',
      },
    },
    hasError: {
      control: {
        type: 'boolean',
      },
    },
    variant: {
      control: {
        type: 'inline-radio',
        options: ['default', 'overlay'],
      },
    },
  },
}

type Props = {
  name: string
  ariaLabel: string
  selected: boolean
  firstOptionForceChecked: boolean
  onChange: (selected: string[]) => void
  disabled?: boolean
  readonly?: boolean
  hasError?: boolean
  variant?: 'default' | 'overlay'
}

const StyledMultiSelectGroup = styled(MultiSelectGroup)`
  display: grid;
  grid-template-columns: 1fr;
  gap: 8px;
`

const Template: Story<Props> = ({
  name,
  ariaLabel,
  selected,
  firstOptionForceChecked,
  onChange,
  disabled,
  readonly,
  hasError,
  variant,
}) => {
  return (
    <StyledMultiSelectGroup
      {...{
        name,
        ariaLabel,
        onChange,
        disabled,
        readonly,
        hasError,
      }}
      className={''}
      selected={selected ? ['選択肢1', '選択肢3'] : []}
    >
      {[1, 2, 3, 4].map((idx) => (
        <MultiSelect
          value={`選択肢${idx}`}
          forceChecked={firstOptionForceChecked && idx === 1}
          variant={variant}
          key={idx}
        >
          選択肢{idx}
        </MultiSelect>
      ))}
    </StyledMultiSelectGroup>
  )
}

export const Default = Template.bind({})
Default.args = {
  name: '',
  ariaLabel: '',
  selected: true,
  firstOptionForceChecked: false,
  disabled: false,
  readonly: false,
  hasError: false,
  variant: 'default',
  // eslint-disable-next-line no-console
  onChange: (selected) => console.log(selected),
}

type PlaygroundProps = {
  name: string
  ariaLabel: string
  disabled?: boolean
  readonly?: boolean
  hasError?: boolean
  variant?: 'default' | 'overlay'
}

export const Playground: Story<PlaygroundProps> = (props) => {
  const [selected, setSelected] = useState<string[]>([])

  return (
    <StyledMultiSelectGroup
      {...props}
      selected={selected}
      onChange={setSelected}
    >
      {[1, 2, 3, 4].map((idx) => (
        <MultiSelect value={`選択肢${idx}`} variant={props.variant} key={idx}>
          選択肢{idx}
        </MultiSelect>
      ))}
    </StyledMultiSelectGroup>
  )
}
Playground.args = {
  name: 'defaultName',
  ariaLabel: '',
  disabled: false,
  readonly: false,
  hasError: false,
  variant: 'default',
}
