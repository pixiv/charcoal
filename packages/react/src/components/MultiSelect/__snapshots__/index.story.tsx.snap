// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots MultiSelect Default 1`] = `
.c1 {
  display: grid;
  grid-template-columns: auto 1fr;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  cursor: pointer;
  gap: 4px;
}

.c1:disabled,
.c1[aria-disabled]:not([aria-disabled='false']) {
  opacity: 0.32;
  cursor: default;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 14px;
  line-height: 22px;
  display: flow-root;
  color: var(--charcoal-text2);
}

.c4::before {
  display: block;
  width: 0;
  height: 0;
  content: '';
  margin-top: -4px;
}

.c4::after {
  display: block;
  width: 0;
  height: 0;
  content: '';
  margin-bottom: -4px;
}

.c2[type='checkbox'] {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  display: block;
  width: 20px;
  height: 20px;
  margin: 0;
  background-color: var(--charcoal-text3);
  border-radius: 999999px;
  -webkit-transition: 0.2s background-color,0.2s box-shadow;
  transition: 0.2s background-color,0.2s box-shadow;
}

.c2[type='checkbox']:checked {
  background-color: var(--charcoal-brand);
}

.c2[type='checkbox']:checked:hover:not(:disabled):not([aria-disabled]),
.c2[type='checkbox']:checked:hover[aria-disabled='false'] {
  background-color: var(--charcoal-brand-hover);
}

.c2[type='checkbox']:checked:active:not(:disabled):not([aria-disabled]),
.c2[type='checkbox']:checked:active[aria-disabled='false'] {
  background-color: var(--charcoal-brand-press);
}

.c2[type='checkbox']:hover:not(:disabled):not([aria-disabled]),
.c2[type='checkbox']:hover[aria-disabled='false'] {
  background-color: var(--charcoal-text3-hover);
}

.c2[type='checkbox']:active:not(:disabled):not([aria-disabled]),
.c2[type='checkbox']:active[aria-disabled='false'] {
  background-color: var(--charcoal-text3-press);
}

.c3 {
  position: absolute;
  top: -2px;
  left: -2px;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 24px;
  height: 24px;
  border-radius: 999999px;
  color: var(--charcoal-text5);
  -webkit-transition: 0.2s box-shadow;
  transition: 0.2s box-shadow;
}

.c0 {
  display: grid;
  grid-template-columns: 1fr;
  gap: 8px;
}

<div
  data-dark={false}
>
  <div
    aria-label=""
    className="c0"
    data-testid="SelectGroup"
  >
    <label
      aria-disabled={false}
      className="c1"
    >
      <input
        aria-invalid={false}
        checked={true}
        className="c2"
        disabled={false}
        name=""
        onChange={[Function]}
        type="checkbox"
        value="選択肢1"
      />
      <div
        aria-hidden={true}
        className="c3"
      >
        <pixiv-icon
          name="24/Check"
          unsafe-non-guideline-scale={0.6666666666666666}
        />
      </div>
      <div
        className="c4"
      >
        選択肢
        1
      </div>
    </label>
    <label
      aria-disabled={false}
      className="c1"
    >
      <input
        aria-invalid={false}
        checked={false}
        className="c2"
        disabled={false}
        name=""
        onChange={[Function]}
        type="checkbox"
        value="選択肢2"
      />
      <div
        aria-hidden={true}
        className="c3"
      >
        <pixiv-icon
          name="24/Check"
          unsafe-non-guideline-scale={0.6666666666666666}
        />
      </div>
      <div
        className="c4"
      >
        選択肢
        2
      </div>
    </label>
    <label
      aria-disabled={false}
      className="c1"
    >
      <input
        aria-invalid={false}
        checked={true}
        className="c2"
        disabled={false}
        name=""
        onChange={[Function]}
        type="checkbox"
        value="選択肢3"
      />
      <div
        aria-hidden={true}
        className="c3"
      >
        <pixiv-icon
          name="24/Check"
          unsafe-non-guideline-scale={0.6666666666666666}
        />
      </div>
      <div
        className="c4"
      >
        選択肢
        3
      </div>
    </label>
    <label
      aria-disabled={false}
      className="c1"
    >
      <input
        aria-invalid={false}
        checked={false}
        className="c2"
        disabled={false}
        name=""
        onChange={[Function]}
        type="checkbox"
        value="選択肢4"
      />
      <div
        aria-hidden={true}
        className="c3"
      >
        <pixiv-icon
          name="24/Check"
          unsafe-non-guideline-scale={0.6666666666666666}
        />
      </div>
      <div
        className="c4"
      >
        選択肢
        4
      </div>
    </label>
  </div>
</div>
`;

exports[`Storyshots MultiSelect Invalid 1`] = `
.c1 {
  display: grid;
  grid-template-columns: auto 1fr;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  cursor: pointer;
  gap: 4px;
}

.c1:disabled,
.c1[aria-disabled]:not([aria-disabled='false']) {
  opacity: 0.32;
  cursor: default;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 14px;
  line-height: 22px;
  display: flow-root;
  color: var(--charcoal-text2);
}

.c4::before {
  display: block;
  width: 0;
  height: 0;
  content: '';
  margin-top: -4px;
}

.c4::after {
  display: block;
  width: 0;
  height: 0;
  content: '';
  margin-bottom: -4px;
}

.c2[type='checkbox'] {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  display: block;
  width: 20px;
  height: 20px;
  margin: 0;
  background-color: var(--charcoal-text3);
  border-radius: 999999px;
  -webkit-transition: 0.2s background-color,0.2s box-shadow;
  transition: 0.2s background-color,0.2s box-shadow;
}

.c2[type='checkbox']:checked {
  background-color: var(--charcoal-brand);
}

.c2[type='checkbox']:checked:hover:not(:disabled):not([aria-disabled]),
.c2[type='checkbox']:checked:hover[aria-disabled='false'] {
  background-color: var(--charcoal-brand-hover);
}

.c2[type='checkbox']:checked:active:not(:disabled):not([aria-disabled]),
.c2[type='checkbox']:checked:active[aria-disabled='false'] {
  background-color: var(--charcoal-brand-press);
}

.c2[type='checkbox']:hover:not(:disabled):not([aria-disabled]),
.c2[type='checkbox']:hover[aria-disabled='false'] {
  background-color: var(--charcoal-text3-hover);
}

.c2[type='checkbox']:active:not(:disabled):not([aria-disabled]),
.c2[type='checkbox']:active[aria-disabled='false'] {
  background-color: var(--charcoal-text3-press);
}

.c2[type='checkbox']:not(:disabled):not([aria-disabled]),
.c2[type='checkbox'][aria-disabled='false'] {
  box-shadow: 0 0 0 4px rgba(255,43,0,0.32);
}

.c3 {
  position: absolute;
  top: -2px;
  left: -2px;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 24px;
  height: 24px;
  border-radius: 999999px;
  color: var(--charcoal-text5);
  -webkit-transition: 0.2s box-shadow;
  transition: 0.2s box-shadow;
}

.c0 {
  display: grid;
  grid-template-columns: 1fr;
  gap: 8px;
}

<div
  data-dark={false}
>
  <div
    aria-label=""
    className="c0"
    data-testid="SelectGroup"
  >
    <label
      aria-disabled={false}
      className="c1"
    >
      <input
        aria-invalid={true}
        checked={false}
        className="c2"
        disabled={false}
        name="defaultName"
        onChange={[Function]}
        type="checkbox"
        value="選択肢1"
      />
      <div
        aria-hidden={true}
        className="c3"
      >
        <pixiv-icon
          name="24/Check"
          unsafe-non-guideline-scale={0.6666666666666666}
        />
      </div>
      <div
        className="c4"
      >
        選択肢
        1
      </div>
    </label>
    <label
      aria-disabled={false}
      className="c1"
    >
      <input
        aria-invalid={true}
        checked={false}
        className="c2"
        disabled={false}
        name="defaultName"
        onChange={[Function]}
        type="checkbox"
        value="選択肢2"
      />
      <div
        aria-hidden={true}
        className="c3"
      >
        <pixiv-icon
          name="24/Check"
          unsafe-non-guideline-scale={0.6666666666666666}
        />
      </div>
      <div
        className="c4"
      >
        選択肢
        2
      </div>
    </label>
    <label
      aria-disabled={false}
      className="c1"
    >
      <input
        aria-invalid={true}
        checked={false}
        className="c2"
        disabled={false}
        name="defaultName"
        onChange={[Function]}
        type="checkbox"
        value="選択肢3"
      />
      <div
        aria-hidden={true}
        className="c3"
      >
        <pixiv-icon
          name="24/Check"
          unsafe-non-guideline-scale={0.6666666666666666}
        />
      </div>
      <div
        className="c4"
      >
        選択肢
        3
      </div>
    </label>
    <label
      aria-disabled={false}
      className="c1"
    >
      <input
        aria-invalid={true}
        checked={false}
        className="c2"
        disabled={false}
        name="defaultName"
        onChange={[Function]}
        type="checkbox"
        value="選択肢4"
      />
      <div
        aria-hidden={true}
        className="c3"
      >
        <pixiv-icon
          name="24/Check"
          unsafe-non-guideline-scale={0.6666666666666666}
        />
      </div>
      <div
        className="c4"
      >
        選択肢
        4
      </div>
    </label>
  </div>
</div>
`;

exports[`Storyshots MultiSelect Overlay 1`] = `
.c1 {
  display: grid;
  grid-template-columns: auto 1fr;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  cursor: pointer;
  gap: 4px;
}

.c1:disabled,
.c1[aria-disabled]:not([aria-disabled='false']) {
  opacity: 0.32;
  cursor: default;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 14px;
  line-height: 22px;
  display: flow-root;
  color: var(--charcoal-text2);
}

.c4::before {
  display: block;
  width: 0;
  height: 0;
  content: '';
  margin-top: -4px;
}

.c4::after {
  display: block;
  width: 0;
  height: 0;
  content: '';
  margin-bottom: -4px;
}

.c2[type='checkbox'] {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  display: block;
  width: 20px;
  height: 20px;
  margin: 0;
  background-color: var(--charcoal-text3);
  border-radius: 999999px;
  -webkit-transition: 0.2s background-color,0.2s box-shadow;
  transition: 0.2s background-color,0.2s box-shadow;
  background-color: var(--charcoal-surface4);
}

.c2[type='checkbox']:checked {
  background-color: var(--charcoal-brand);
}

.c2[type='checkbox']:checked:hover:not(:disabled):not([aria-disabled]),
.c2[type='checkbox']:checked:hover[aria-disabled='false'] {
  background-color: var(--charcoal-brand-hover);
}

.c2[type='checkbox']:checked:active:not(:disabled):not([aria-disabled]),
.c2[type='checkbox']:checked:active[aria-disabled='false'] {
  background-color: var(--charcoal-brand-press);
}

.c2[type='checkbox']:hover:not(:disabled):not([aria-disabled]),
.c2[type='checkbox']:hover[aria-disabled='false'] {
  background-color: var(--charcoal-text3-hover);
}

.c2[type='checkbox']:active:not(:disabled):not([aria-disabled]),
.c2[type='checkbox']:active[aria-disabled='false'] {
  background-color: var(--charcoal-text3-press);
}

.c3 {
  position: absolute;
  top: -2px;
  left: -2px;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 24px;
  height: 24px;
  border-radius: 999999px;
  color: var(--charcoal-text5);
  -webkit-transition: 0.2s box-shadow;
  transition: 0.2s box-shadow;
  border-color: var(--charcoal-text5);
  border-width: 2px;
  border-style: solid;
}

.c0 {
  display: grid;
  grid-template-columns: 1fr;
  gap: 8px;
}

<div
  data-dark={false}
>
  <div
    aria-label=""
    className="c0"
    data-testid="SelectGroup"
  >
    <label
      aria-disabled={false}
      className="c1"
    >
      <input
        aria-invalid={false}
        checked={false}
        className="c2"
        disabled={false}
        name="defaultName"
        onChange={[Function]}
        type="checkbox"
        value="選択肢1"
      />
      <div
        aria-hidden={true}
        className="c3"
      >
        <pixiv-icon
          name="24/Check"
          unsafe-non-guideline-scale={0.6666666666666666}
        />
      </div>
      <div
        className="c4"
      >
        選択肢
        1
      </div>
    </label>
    <label
      aria-disabled={false}
      className="c1"
    >
      <input
        aria-invalid={false}
        checked={false}
        className="c2"
        disabled={false}
        name="defaultName"
        onChange={[Function]}
        type="checkbox"
        value="選択肢2"
      />
      <div
        aria-hidden={true}
        className="c3"
      >
        <pixiv-icon
          name="24/Check"
          unsafe-non-guideline-scale={0.6666666666666666}
        />
      </div>
      <div
        className="c4"
      >
        選択肢
        2
      </div>
    </label>
    <label
      aria-disabled={false}
      className="c1"
    >
      <input
        aria-invalid={false}
        checked={false}
        className="c2"
        disabled={false}
        name="defaultName"
        onChange={[Function]}
        type="checkbox"
        value="選択肢3"
      />
      <div
        aria-hidden={true}
        className="c3"
      >
        <pixiv-icon
          name="24/Check"
          unsafe-non-guideline-scale={0.6666666666666666}
        />
      </div>
      <div
        className="c4"
      >
        選択肢
        3
      </div>
    </label>
    <label
      aria-disabled={false}
      className="c1"
    >
      <input
        aria-invalid={false}
        checked={false}
        className="c2"
        disabled={false}
        name="defaultName"
        onChange={[Function]}
        type="checkbox"
        value="選択肢4"
      />
      <div
        aria-hidden={true}
        className="c3"
      >
        <pixiv-icon
          name="24/Check"
          unsafe-non-guideline-scale={0.6666666666666666}
        />
      </div>
      <div
        className="c4"
      >
        選択肢
        4
      </div>
    </label>
  </div>
</div>
`;

exports[`Storyshots MultiSelect Playground 1`] = `
.c1 {
  display: grid;
  grid-template-columns: auto 1fr;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  cursor: pointer;
  gap: 4px;
}

.c1:disabled,
.c1[aria-disabled]:not([aria-disabled='false']) {
  opacity: 0.32;
  cursor: default;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 14px;
  line-height: 22px;
  display: flow-root;
  color: var(--charcoal-text2);
}

.c4::before {
  display: block;
  width: 0;
  height: 0;
  content: '';
  margin-top: -4px;
}

.c4::after {
  display: block;
  width: 0;
  height: 0;
  content: '';
  margin-bottom: -4px;
}

.c2[type='checkbox'] {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  display: block;
  width: 20px;
  height: 20px;
  margin: 0;
  background-color: var(--charcoal-text3);
  border-radius: 999999px;
  -webkit-transition: 0.2s background-color,0.2s box-shadow;
  transition: 0.2s background-color,0.2s box-shadow;
}

.c2[type='checkbox']:checked {
  background-color: var(--charcoal-brand);
}

.c2[type='checkbox']:checked:hover:not(:disabled):not([aria-disabled]),
.c2[type='checkbox']:checked:hover[aria-disabled='false'] {
  background-color: var(--charcoal-brand-hover);
}

.c2[type='checkbox']:checked:active:not(:disabled):not([aria-disabled]),
.c2[type='checkbox']:checked:active[aria-disabled='false'] {
  background-color: var(--charcoal-brand-press);
}

.c2[type='checkbox']:hover:not(:disabled):not([aria-disabled]),
.c2[type='checkbox']:hover[aria-disabled='false'] {
  background-color: var(--charcoal-text3-hover);
}

.c2[type='checkbox']:active:not(:disabled):not([aria-disabled]),
.c2[type='checkbox']:active[aria-disabled='false'] {
  background-color: var(--charcoal-text3-press);
}

.c3 {
  position: absolute;
  top: -2px;
  left: -2px;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 24px;
  height: 24px;
  border-radius: 999999px;
  color: var(--charcoal-text5);
  -webkit-transition: 0.2s box-shadow;
  transition: 0.2s box-shadow;
}

.c0 {
  display: grid;
  grid-template-columns: 1fr;
  gap: 8px;
}

<div
  data-dark={false}
>
  <div
    aria-label=""
    className="c0"
    data-testid="SelectGroup"
  >
    <label
      aria-disabled={false}
      className="c1"
    >
      <input
        aria-invalid={false}
        checked={false}
        className="c2"
        disabled={false}
        name="defaultName"
        onChange={[Function]}
        type="checkbox"
        value="選択肢1"
      />
      <div
        aria-hidden={true}
        className="c3"
      >
        <pixiv-icon
          name="24/Check"
          unsafe-non-guideline-scale={0.6666666666666666}
        />
      </div>
      <div
        className="c4"
      >
        選択肢
        1
      </div>
    </label>
    <label
      aria-disabled={false}
      className="c1"
    >
      <input
        aria-invalid={false}
        checked={false}
        className="c2"
        disabled={false}
        name="defaultName"
        onChange={[Function]}
        type="checkbox"
        value="選択肢2"
      />
      <div
        aria-hidden={true}
        className="c3"
      >
        <pixiv-icon
          name="24/Check"
          unsafe-non-guideline-scale={0.6666666666666666}
        />
      </div>
      <div
        className="c4"
      >
        選択肢
        2
      </div>
    </label>
    <label
      aria-disabled={false}
      className="c1"
    >
      <input
        aria-invalid={false}
        checked={false}
        className="c2"
        disabled={false}
        name="defaultName"
        onChange={[Function]}
        type="checkbox"
        value="選択肢3"
      />
      <div
        aria-hidden={true}
        className="c3"
      >
        <pixiv-icon
          name="24/Check"
          unsafe-non-guideline-scale={0.6666666666666666}
        />
      </div>
      <div
        className="c4"
      >
        選択肢
        3
      </div>
    </label>
    <label
      aria-disabled={false}
      className="c1"
    >
      <input
        aria-invalid={false}
        checked={false}
        className="c2"
        disabled={false}
        name="defaultName"
        onChange={[Function]}
        type="checkbox"
        value="選択肢4"
      />
      <div
        aria-hidden={true}
        className="c3"
      >
        <pixiv-icon
          name="24/Check"
          unsafe-non-guideline-scale={0.6666666666666666}
        />
      </div>
      <div
        className="c4"
      >
        選択肢
        4
      </div>
    </label>
  </div>
</div>
`;
